### PR TITLE
correct masking: switch for each example in a batch

### PR DESCRIPTION
**Pull request by jeffzhengye**
[  # input 0 in a batch
               [0], [1], [2], [3]
               ],
            [  # input 1 in a batch
               [0], [0], [1], [2]
               ]
when we compute 'switch' for the second example ([1], [0]), the switch will be '1'.  Obvious, for input 1, switch should be 0. 
Actually, switch should not be a scalar, instead it should be a vector (1, 0). 

@fchollet Another huge bug in RNN I mentioned at https://github.com/fchollet/keras/issues/856#issuecomment-157771075. How do you think? Any plan to fix this?

### CODE REVIEW

**Comment by fchollet** [comment]
Could you add a simple unit test (just a few lines in `test_recurrent.py`) checking the correctness of the masking behavior?

**Comment by jeffzhengye** [comment]
@fchollet not sure how to fix the test in test_recurrent.py. But still added a test to verify the correctness.

**Comment by jeffzhengye** [comment]
Anyone can help with the test? @fchollet

**Comment by fchollet** [comment]
I think you just need to:
- not import modules that aren't part of the Travis-installed stack
- use pytest, not unittest
- incorporate your test into `tests/keras/layers/test_recurrent.py`, as a standalone function for instance.

Existing test functions in `tests/keras` should give you an idea of how to proceed.

**Comment by fchollet** [comment]
For now you can just remove the test script from the PR. I will write tests for masking tomorrow then merge.

**Comment by jeffzhengye** [comment]
@fchollet OK. Thanks. 
can you also check the bug in RNN I mentioned at #856 (comment). How do you think? Any plan to fix this?

**Comment by fchollet** [comment]
> Any plan to fix this?

Maybe, I agree that it's an issue. But it isn't clear what the best course of action would be. For now the simplest thing to do is to make sure that your padding side (which you can configure) matches your RNN direction.

**Comment by jeffzhengye** [comment]
No, @fchollet, That's a bug, not an issue, if you want to use bi-directional rnn as in imdb_bidirectional_lstm.py. 

For a simple example,  say we use simple RNN to mimic addition with input [[0], [1], [2], [3] ], forward rnn will output [[0], [1], [3], [6] ], and baward will output [[3], [5], [6], [0]]

If return_sequence=False, bidirectional will output [[6, 0]], instead of [[6, 6]], which is incorrect. 
If we claim not supporting bi-directional rnn, we should make it explicit and remove this example.

**Comment by farizrahman4u** [comment]
@jeffzhengye NO!!!!

Backward will output [[3], [5], [6], [6]]

**Comment by jeffzhengye** [comment]
@farizrahman4u No. Here, 0 indicates mask.

**Comment by jeffzhengye** [comment]
@farizrahman4u If there is no masking, everything is fine indeed.

**Comment by farizrahman4u** [comment]
Oh! Ok I get it.  When you wrote 0,1,2,3  I thought zero was part of input.

**Comment by jeffzhengye** [comment]
@farizrahman4u please help to confirm this.

**Comment by farizrahman4u** [comment]
But isnt this taken care of by the T.switch ?

**Comment by jeffzhengye** [comment]
No. T.switch can only make sure padded time step returns 0.
But there is no way to know which time step is the actual last step when
go_backwards=True, unless we switch the padding when go_backward.

Still the same example, [[0], [1], [2], [3]], when we go backward, the
padded zero should be on the right (which is [[1], [2], [3], [0]]),
If we can do a padding switch to get that, it should be correct when
return_sequence=False.

But It will cause problem when return_sequence=True, because when to
concatenate forward and backward for the corresponding time step, the
alignment has already messed up after padding switch.
We can consider to create a BidirectionRnn class to take care of these, but
I only have an awkward implementation. I hope someone could think out an
nice one.

Please note when considering this problem, I always assume it is in case of
bi-directional RNN.


On Tue, Dec 15, 2015 at 2:54 PM, Fariz Rahman <notifications@github.com>
wrote:

> But isnt this taken care of by the T.switch ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fchollet/keras/pull/1258#issuecomment-164876418>.
>



-- 
Jeff Ye.

**Comment by farizrahman4u** [comment]
Is your crude implementation on github? Or just post the main stuff here. I will take care of the rest.

**Comment by jeffzhengye** [comment]
Yes, I can post it here if you confirm this is really a bug. I personally
believe so.

On Tue, Dec 15, 2015 at 3:27 PM, Fariz Rahman <notifications@github.com>
wrote:

> Is your crude implementation on github? Or just post the main stuff here.
> I will take care of the rest.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fchollet/keras/pull/1258#issuecomment-164886991>.
>



-- 
Jeff Ye.

**Comment by farizrahman4u** [comment]
Yes. Post it.

**Comment by jeffzhengye** [comment]
class PaddingSwitch(MaskedLayer):
    def __init__(self, ndim, left_to_right, do_infer_mask=True):
        """
            @:param left_to_right: True if switching from left padding to right padding;
                False if switching from right padding to left padding
            @:param do_infer_mask: if True then infer the mask from the input
        """
        self.ndim = ndim
        self.left_to_right = left_to_right
        self.do_infer_mask = do_infer_mask
        self.params = []

    def infer_mask(self, train):  # X ~ n_sample x n_time x ...
        X = self.get_input(train)
        n_sample, n_time = X.shape[:2]
        if self.ndim == 2:
            mask = T.ones((n_sample, n_time)) * (1 - T.eq(X, 0))
        else:
            mask = T.ones((n_sample, n_time)) * (1 - T.all(T.eq(X, 0), axis=2))
        return mask

    def get_index(self, train=False):
        if self.do_infer_mask or not hasattr(self, 'previous') or not self.previous.get_output_mask():
            X = self.get_input(train)
            mask = self.infer_mask(X)
        else:
            mask = self.get_input_mask(train)  # n_sample x n_time
        n_sample, n_time = mask.shape
        if self.left_to_right:
            indexs = n_time - T.sum(mask, axis=1)
        else:
            indexs = T.sum(mask, axis=1)
        return indexs, mask

    def get_output_mask(self, train=False):
        X = self.get_input(train)  # n_sample x n_time
        assert X.ndim == self.ndim
        indexs, mask = self.get_index(train)
        mask, updates = theano.scan(lambda x, i: T.roll(x, -i, 0),
                                    sequences=[mask, indexs])
        return mask.astype('int32')

    def get_output(self, train=False):
        X = self.get_input(train)
        assert X.ndim == self.ndim
        indexs, mask = self.get_index(train)
        X, updates = theano.scan(lambda x, i: T.roll(x, -i, 0),
                                 sequences=[X, indexs])
        return X



class Inverse(MaskedLayer):
    def __init__(self):
        self.params = []

    def get_output(self, train=False):
        X = self.get_input(train)
        return X[:, ::-1]

    def get_output_mask(self, train=False):
        mask = self.get_input_mask(train)
        if mask == None:
            return mask
        else:
            return mask[:, ::-1]




class Bidirectional(Recurrent):
    ndim = 3

    def __init__(self, is_left_padding=True, forward=None, backward=None,
                 forward_conf=None, backward_conf=None, **kwargs):
        assert forward is not None or forward_conf is not None, "Must provide a forward RNN or a forward configuration"
        assert backward is not None or backward_conf is not None, "Must provide a backward RNN or a backward configuration"
        assert not hasattr(forward, 'go_backwards') or not forward.go_backwards, "forward RNN should not go backwards"
        # assert not hasattr(backward,
        #                    'go_backwards') or not backward.go_backwards, "backward RNN should not go backwards"
        super(Bidirectional, self).__init__(**kwargs)
        if forward is not None:
            self.forward = forward
        else:
            # Must import inside the function, because in order to support loading
            # we must import this module inside layer_utils.
            from keras.utils.layer_utils import container_from_config
            self.forward = container_from_config(forward_conf)
        if backward is not None:
            self.backward = backward
        else:
            from keras.utils.layer_utils import container_from_config
            self.backward = container_from_config(backward_conf)
        self.forward.return_sequences = self.return_sequences
        self.backward.return_sequences = True
        self.backward.go_backwards = True
        self.is_left_padding = is_left_padding
        self.output_dim = self.forward.output_dim + self.backward.output_dim

    def set_previous(self, layer, connection_map={}):
        assert self.nb_input == layer.nb_output == 1, "Cannot connect layers: input count and output count should be 1."
        if hasattr(self, 'input_ndim'):
            assert self.input_ndim == len(layer.output_shape), "Incompatible shapes: layer expected input with ndim=" + \
                                                               str(
                                                                   self.input_ndim) + " but previous layer has output_shape " + str(
                layer.output_shape)
        if self.is_left_padding:  # 0,0,1,2,3
            self.forward.set_previous(layer, connection_map)  # 0,0,1,3,6
            self.backward.set_previous(layer, connection_map)  # 3,5,6,0,0
            self.padding_switch = PaddingSwitch(self.ndim, False, False)  # left -> right, 0,0,3,5,6
            self.padding_switch.set_previous(self.backward)  # 0,0,3,5,6

            self.inverse2 = Inverse()  # this is for return_sequence=True
            self.inverse2.set_previous(self.backward)  # 0,0,6,5,3
        else:  # 1,2,3,0,0
            # implementation for right padding
            raise NotImplementedError
        self.params = self.forward.params + self.backward.params
        super(Bidirectional, self).set_previous(layer, connection_map={})

    @property
    def output_shape(self):
        input_shape = self.input_shape
        if self.return_sequences:
            return (input_shape[0], input_shape[1], self.output_dim)
        else:
            return (input_shape[0], self.output_dim)

    def _get_customize_output_shape(self, return_sequences):
        input_shape = self.input_shape
        if return_sequences:
            return (input_shape[0], input_shape[1], self.output_dim)
        else:
            return (input_shape[0], self.output_dim)

    def get_output(self, train=False):
        return self.get_customized_output(self.return_sequences, train)

    def get_customized_output(self, return_sequences, train=False):
        if not return_sequences:
            return T.concatenate([self.forward.get_output(train), self.padding_switch.get_output(train)[:, -1]],
                                 axis=1)
        else:
            return T.concatenate([self.forward.get_output(train), self.inverse2.get_output(train)], axis=2)

    def get_output_mask(self, train=False):
        return self.get_input_mask(train) if self.return_sequences else None

**Comment by farizrahman4u** [comment]
Too theanoish. I will look into it.

**Comment by farizrahman4u** [comment]
@jeffzhengye Check this out #1282
